table/column.go

```diff
@@ -315,21 +315,23 @@ func CastValue(ctx sessionctx.Context, val types.Datum, col *model.ColumnInfo, r
 	str := casted.GetString()
 	utf8Charset := col.Charset == mysql.UTF8Charset
 	doMB4CharCheck := utf8Charset && config.GetGlobalConfig().CheckMb4ValueInUTF8
-	for i, w := 0, 0; i < len(str); i += w {
-		runeValue, width := utf8.DecodeRuneInString(str[i:])
-		if runeValue == utf8.RuneError {
-			if strings.HasPrefix(str[i:], string(utf8.RuneError)) {
-				w = width
-				continue
+	if !utf8.ValidString(str) {
```

Comment:
This is the only line of changes...

Reply:
If doMB4CharCheck is true and charset is utf8, a char with 4-bytes will insert successfully. But it shouldn't.

Reply:
fixed